src/uct/ib/base/ib_iface.c

```diff
@@ -1173,6 +1174,88 @@ static void uct_ib_iface_set_path_mtu(uct_ib_iface_t *iface,
     }
 }
 
+const char*
+uct_ib_iface_sl_mask_str(uint16_t sl_mask, ucs_string_buffer_t *str_b)
+{
+    uint8_t sl;
+
+    if (sl_mask == 0) {
+        ucs_string_buffer_appendf(str_b, "<none>");
+        goto out;
+    }
+
+    ucs_for_each_bit(sl, sl_mask) {
+        ucs_assert(sl < 16);
+
+        if (ucs_string_buffer_length(str_b) == 0) {
+            ucs_string_buffer_appendf(str_b, "%u", sl);
+        } else {
+            ucs_string_buffer_appendf(str_b, ", %u", sl);
+        }
+    }
+
+out:
+    return ucs_string_buffer_cstr(str_b);
+}
+
+ucs_status_t
+uct_ib_iface_select_sl(const uct_ib_iface_config_t *ib_config,
+                       ucs_ternary_value_t ar_enable, uint16_t ooo_sl_mask,
+                       int ooo_sl_mask_is_detected,
+                       const char *dev_port_name, uint8_t *sl_p)
+{
+    ucs_status_t status = UCS_OK;
+    ucs_string_buffer_t sl_mask_str;
+    char sl_str[8];
+    uint8_t sl, sl_from_mask;
+
+    ucs_string_buffer_init(&sl_mask_str);
+
+    /* Select the default  */
+    ucs_assert((ib_config->sl <= UINT8_MAX) ||
+               (ib_config->sl == UCS_ULUNITS_AUTO));
+    sl = (ib_config->sl == UCS_ULUNITS_AUTO) ? 0 : (uint8_t)ib_config->sl;
+
+    if (ar_enable == UCS_NO) {
+        goto out;
+    }
+
+    if (ib_config->sl == UCS_ULUNITS_AUTO) {
+        /* Select the first available SL with AR support, since a specific
+         * SL was not requested by user */
+        sl_from_mask = ucs_ffs64_safe(ooo_sl_mask);
+        if (sl_from_mask != 64) {
+            sl = sl_from_mask;
+            goto out;
+        } else if (ar_enable == UCS_YES) {
+            goto err;
+        }
+    } else if (!(ooo_sl_mask & UCS_BIT(ib_config->sl)) &&
+               (ar_enable == UCS_YES)) {
+        goto err;
+    }
+
+out:
+    *sl_p = sl;
+    ucs_debug("SL=%u was selected, SLs with AR support = { %s } on %s",
+              sl, uct_ib_iface_sl_mask_str(ooo_sl_mask, &sl_mask_str),
+              dev_port_name);
+out_str_buf_clean:
+    ucs_string_buffer_cleanup(&sl_mask_str);
+    return status;
+
+err:
+    ucs_config_sprintf_ulong(sl_str, sizeof(sl_str), &ib_config->sl, NULL);
```

Comment:
why not printing SL as int?

Reply:
since it could be `"auto"`

Reply:
which is also `int`.  Why using ` ucs_config_sprintf_ulong` is better?

Reply:
> which is also `int`. Why using ` ucs_config_sprintf_ulong` is better?

right, but it will be printed as `UINT_MAX - 1` value

Reply:
whats is the difference between printing it as int and invoking `ucs_config_sprintf_ulong`?

Reply:
i found this one `ucs_config_sprintf_ulunits`. Maybe you were going to use it instead?

Reply:
> i found this one `ucs_config_sprintf_ulunits`. Maybe you were going to use it instead?

yep, I meant it
fixed
```
[     INFO ] < AR support is requested for SL=auto, but could not select SL with AR support on ib_dev:1, SLs with AR support = { <none> } >
[     INFO ] < AR support is requested for SL=8, but could not select SL with AR support on ib_dev:1, SLs with AR support = { <none> } >
[     INFO ] < AR support is requested for SL=8, but could not select SL with AR support on ib_dev:1, SLs with AR support = { 0 } >
[     INFO ] < AR support is requested for SL=8, but could not select SL with AR support on ib_dev:1, SLs with AR support = { 0, 4, 11 } >
```